Update ebuild syntax highlighting

xxc3nsoredxx opened the pull request:
Added some missing keywords, dolib* and newlib* switched to regex match
to properly highlight them, added a testbench file for testing syntax
highlighting.

Signed-off-by: Oskari Pirhonen <xxc3ncoredxx@gmail.com>

xxc3nsoredxx committed:
Update ebuild syntax highlighting

Added some missing keywords, dolib* and newlib* switched to regex match
to properly highlight them, added a testbench file for testing syntax
highlighting.

Signed-off-by: Oskari Pirhonen <xxc3ncoredxx@gmail.com>

mgorny commented on test/test.ebuild:
```diff
@@ -0,0 +1,200 @@
+# Copyright 2021 Gentoo Authors
```
I don't think this file is being actually used anywhere, is it?

xxc3nsoredxx replied:
I believe that was added automatically when I ran `vim test.ebuild` to create the file.

mgorny replied:
I meant the whole file.

xxc3nsoredxx replied:
OK, I've removed it.

mgorny commented on syntax/ebuild.vim:
```diff
@@ -31,13 +31,16 @@ syn match bkshFunction	"^\s*\<\h[0-9a-zA-Z_\-\.]*\>\s*()"	skipwhite skipnl conta
 syn keyword EbuildCoreKeyword hasv usev usex elog eapply eapply_user
 syn keyword EbuildCoreKeyword einstalldocs in_iuse get_libdir
+syn keyword EbuildCoreKeyword assert nonfatal docompress dostrip einfon eqawarn
+syn match EbuildCoreKeyword "\<dolib\(\.a\|\.so\)\?\_s"
```
Could you explain why this needed to be done? In a comment, preferably.

xxc3nsoredxx replied:
Added.

mgorny replied:
Did you try escaping the dot somehow, or maybe reversing the order?

xxc3nsoredxx replied:
I did yes. I tried various combinations of `dolib.a dolib.so dolib`, `dolib dolib\.a dolib\.so`, `dolib\.a dolib\.so dolib`, `'dolib.a' 'dolib.so' dolib`, and `'dolib\.a' 'dolib\.so' dolib`, as well as with double quotes instead of single quotes, and similar variations on their own lines (similarly with `newlib`) before I went with `syn match` instead of `syn keyword`.

xxc3nsoredxx replied:
Your comments prompted me to do a bit more research, and I found a solution I think you'll like better than my previous `match`-based one. The reason `keyword` wasn't highlighting `dolib.*` and `newlib.*` was because '.' wasn't in the list of keyword characters. Appending '.' to `iskeyword` in `ftplugin/ebuild.vim` allowed me to restore the original `keyword` method while also matching `dolib.a` and friends.

mgorny replied:
Great, thanks! However, this makes me wonder if we shouldn't extend that list even further ;-). I mean, right now a thing like literal `dobin/dolib` gets highlighted even though it forms a path rather than ebuild function call.

xxc3nsoredxx replied:
Hmmmm, I see what you mean. I will take a look to see if I can figure something out.

xxc3nsoredxx replied:
As it is right now, I don't think it's possible because `keyword` is purely keyword-based and takes precedence over `match` (which allows more complicated structures). But if I restructured the syntax file, then I think it could be done. Here's a trivial example of what I mean:

```vim
" Match paths
syn match EbuildPath "\f\+\(/\f\+\)*"
syn match EbuildCoreKeyword "^dotest" nextgroup=EbuildPath skipwhite
"... snip ...
hi def link EbuildPath String
```

Using the following test case:
```
dotest a/a
dotest dotest/a "${A}/a"
dotest dotest path/number/2 dotest
```
On the first line, `dotest` is highlighted as an `EbuildCoreKeyword` and `a/a` is highlighted as an `EbuildPath`. On the second line, the first `dotest` is highlighted as an `ECK`, but `dotest/a` is highlighted as an `EP`, and `"${A}/a"` is highlighted however the `sh.vim` file does it. On the third line, the first `dotest` is highlighted as an `ECK`, the second `dotest` is highlighted as an `EP`, as is `path/number/2` and the third `dotest`.

Note: `dotest` was defined just for experimentation.

CyberTailor replied:
Just add `.` to the list of symbols that can form a keyword.

xxc3nsoredxx committed:
Fixes

Removed the automatically generated copyright notice.
Added comment describing why regex match was used instead of keyword.

Signed-off-by: Oskari Pirhonen <xxc3ncoredxx@gmail.com>

xxc3nsoredxx committed:
Remove test file

Signed-off-by: Oskari Pirhonen <xxc3ncoredxx@gmail.com>

xxc3nsoredxx committed:
Restore 'match' to 'keyword'

Signed-off-by: Oskari Pirhonen <xxc3ncoredxx@gmail.com>

xxc3nsoredxx committed:
Context-aware highlighting for path-y keywords

Path-y keywords are keywords that take a path as an argument such as
"dodir" and "dolib".

Signed-off-by: Oskari Pirhonen <xxc3ncoredxx@gmail.com>

xxc3nsoredxx commented on syntax/ebuild.vim:
```diff
@@ -27,17 +29,32 @@ runtime syntax/gentoo-common.vim
 " this. See Gentoo bug 72469.
 syn match bkshFunction	"^\s*\<\h[0-9a-zA-Z_\-\.]*\>\s*()"	skipwhite skipnl contains=bkshFunctionParen
 
+" Match paths (including split across multiple lines)
+syn match MatchPath contains=@shDblQuoteList,shComment "\f*"
+syn match MatchPath contains=@shDblQuoteList,shComment "\\\_s"
+
+" Match the start of a path-y keyword
+syn region EbuildPathKeyword start="^\s*\k\+\s\@=" end="\>" contains=EbuildPathKeywordList
+" Match the path portion
+syn region EbuildPath start="\>\s\@=" skip="\\$" end="$" contains=MatchPath
+
+" Path-y keywords
+syn keyword EbuildPathKeywordList contained unpack dodir into keepdir dobin dosbin doinitd doconfd doenvd
+syn keyword EbuildPathKeywordList contained dolib dolib.a dolib.so doinfo domo insinto exeinto doexe
+syn keyword EbuildPathKeywordList contained docinto
+
```
@mgorny This latest version highlights only the path-y keywords listed if they're detected as having a valid format. So things like `dolib path/to/file` gets detected as `dolib` + arguments whereas a plain `dolib/path/to/file` gets detected as a path (presumably to some executable file being run).

I know I previously said that I didn't think it was possible, but I managed to do it with `keyword` which you're so fond of, and I can see why you are! The way that I've done it, adding a new keyword to the highlighting only involves adding the keyword to `EbuildPathKeywordList`. It keeps it much more maintainable.

There are still a good number of keywords I still have to handle, such as `doman` which takes an optional argument, and (if possible) I want that argument to be highlighted separately from the list of manpages. As well as non-path-y keywords and path-y keywords I may have missed.

mgorny replied:
Sounds like you're up to some cool stuff. Please let me know when it's ready to test/merge.

xxc3nsoredxx replied:
Will do!

msva commented:
ping @xxc3nsoredxx ?

It looks like great work that needs a little more love to be merged.
Are you still up on it?

xxc3nsoredxx commented:
@msva Yes, I am. This has actually been floating around my head on and off with different ideas coming and going. I'm taking my last final tomorrow, and then I'll be done with grad school! Means I can spend more of my free time doing (and catching up on) personal projects instead of school projects.

xxc3nsoredxx commented:
@mgorny @msva update for you: I've started work on a complete overhaul, basing the format on `syntax/sh.vim` as much as possible. It won't be in a usable state for some amount of time, but you can follow my progress over on a separate branch if interested: https://github.com/xxc3nsoredxx/gentoo-syntax/tree/overhaul

Waiting to open a PR with that branch before closing this one.

msva commented:
@xxc3nsoredxx ping? :)

xxc3nsoredxx commented:
I'm sorry! I've been nerdsniped by various things, and I'm not quite sure when I'll next have a chance to circle back around to this :(

But hopefully Soon (TM)
